kernel/src/drivers/keyboard/mod.rs

```diff
@@ -0,0 +1,200 @@
+
+pub mod keymap;
+
+use core::convert::From;
+
+use drivers::ps2::{self, Device, DeviceState};
+use drivers::ps2::io::Ps2Error;
+use drivers::ps2::io::commands::{DeviceCommand, DeviceDataCommand};
+
+/// Interface to a generic keyboard.
+pub trait Keyboard {
+    type Error;
+
+    /// Enables this keyboard, setting it up before use.
+    ///
+    /// # Note
+    ///
+    /// The keyboard should not be accessed while not enabled!
+    fn enable(&mut self) -> Result<(), Self::Error>;
+
+    /// Disables this keyboard, making use unavailable.
+    ///
+    /// # Note
+    ///
+    /// Until `enable` is called again, this keyboard should not be used
+    fn disable(&mut self) -> Result<(), Self::Error>;
+
+    /// Polls the device for state events.
+    /// TODO: This should eventually use interrupts and hold a queue
+    fn read_event(&mut self) -> Result<KeyEvent, Self::Error>;
+
+    /// Returns true if the given keycode is currently being pressed
+    fn pressed(&self, keycode: u8) -> bool;
+}
+
+/// Contains data relating to a key press event.
+#[derive(Copy, Clone, Debug)]
+pub struct KeyEvent {
+    pub keycode: u8,
+    pub char: Option<char>,
+    pub event_type: KeyEventType,
+    pub ctrl: bool,
+    pub alt: bool,
+    pub shift: bool,
+}
+
+/// The type of key event occurring
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum KeyEventType {
+    Make,
+    Break,
+    Repeat,
+}
+
+/// An error for a PS/2 keyboard
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum Ps2KeyboardError {
+    ReadError(Ps2Error),
+    KeyboardDisabled,
+    ScancodeNotSet(u8),
+    ScanningNotEnabled,
+    DataUnavailable,
+}
+
+/// Handles interface to a PS/2 keyboard, if available
+pub struct Ps2Keyboard<'a> {
+    device: &'a mut Device,
+    key_states: [bool; 0xFF],
+}
+
+impl<'a> Ps2Keyboard<'a> {
+    pub fn new(device: &'a mut Device) -> Self {
+        Ps2Keyboard {
+            device,
+            key_states: [false; 0xFF],
+        }
+    }
+
+    /// Reads a single scancode from this
+    fn read_scancode(&self) -> Result<Ps2Scancode, Ps2KeyboardError> {
+        self.check_enabled()?;
+
+        if ps2::io::can_read()? && ps2::io::can_read_keyboard()? {
+            // Check scancode present
+            let scancode = ps2::io::read(&ps2::io::DATA_PORT)?;
+
+            // Check not extended scancode (ignored)
+            if scancode != 0 && scancode != 0xE0 && scancode != 0xE1 {
+                // If not key release (break) code
+                if scancode < 0x80 {
+                    return Ok(Ps2Scancode::new(scancode, true));
+                } else {
+                    return Ok(Ps2Scancode::new(scancode - 0x80, false));
+                }
+            }
+        }
+
+        Err(Ps2KeyboardError::DataUnavailable)
+    }
+
+    /// Checks if this keyboard is enabled, and returns an error if not
+    fn check_enabled(&self) -> Result<(), Ps2KeyboardError> {
```

Comment:
This could probably return a boolean

Reply:
It could, but this method is used where the result is propagated.

Reply:
IMO it doesn't matter, it should still be a boolean. It could be mapped to an error or something.

Reply:
The boolean check is a simple condition, if it is done like that, there is no reason for this method.

Reply:
Then perhaps this method shouldn't exist? IMO it feels... unnatural to do something like this :/

Reply:
How should callers handle this instead?

Reply:
Just by doing `if self.state == enabled` or whatever it is.